app/src/backend/ActivityService.ts

```diff
@@ -21,9 +21,241 @@ import type { Transaction } from "sequelize";
+        } else {
+          emissionsFactor = await db.models.EmissionsFactor.create(
+            {
+              ...gasValue.emissionsFactor,
+              id: randomUUID(),
+              inventoryId: activityValue.inventoryValueId,
+            },
+            { transaction },
+          );
+        }
+      }
+
+      if (gasValue.emissionsFactor && !emissionsFactor) {
+        throw new createHttpError.InternalServerError(
+          "Failed to create an emissions factor",
+        );
+      }
+
+      delete gasValue.emissionsFactor;
+
+      if (gasValue.gasAmount == null) {
+        gasValue.gasAmount =
+          gases.find((gas) => gas.gas === gasValue.gas)?.amount ?? 0n;
+      }
+
+      await db.models.GasValue.upsert(
+        {
+          emissionsFactorId: emissionsFactor?.id ?? undefined,
+          ...gasValue,
+          id: gasValue.id ?? randomUUID(),
+          activityValueId: activityValue.id,
+          inventoryValueId: activityValue?.inventoryValueId,
+        },
+        { transaction },
+      );
+    }
+  }
+
+  public static async updateActivity({
+    id,
+    inventoryValueId,
+    inventoryValueParams,
+    activityValueParams,
+    gasValues,
+    dataSourceParams,
+  }: {
+    id: string;
+    activityValueParams: Omit<ActivityValueAttributes, "id">;
+    inventoryValueId: string | undefined;
+    inventoryValueParams: Omit<InventoryValueAttributes, "id"> | undefined;
+    gasValues: GasValueInput[] | undefined;
+    dataSourceParams: Omit<DataSourceAttributes, "datasourceId"> | undefined;
+  }): Promise<ActivityValue | undefined> {
+    const activityValue = await db.models.ActivityValue.findOne({
+      where: { id },
+    });
+
+    if (!activityValue) {
+      throw new createHttpError.NotFound(
+        `Activity value with ID ${id} not found`,
+      );
+    }
+
+    let datasourceId = activityValue.datasourceId;
+
+    return await db.sequelize?.transaction(
+      async (transaction): Promise<ActivityValue> => {
+        datasourceId = await this.updateDataSource({
+          activityValue,
+          dataSourceParams,
+          transaction,
+        });
+
+        const inventoryValue = await this.updateInventory({
+          activityValue,
+          inventoryValueParams,
+        });
+
+        // update the activity value with new params
+        await activityValue.update({
+          ...activityValueParams,
+          datasourceId,
+          inventoryValueId,
+        });
+
+        // CO2 calculation
+        let { totalCO2e, totalCO2eYears, gases } =
+          await CalculationService.calculateGasAmount(
+            inventoryValue,
+            activityValue,
+            inventoryValue.inputMethodology as string,
+          );
+
+        inventoryValue.co2eq = totalCO2e;
+        inventoryValue.co2eqYears = Math.max(0, totalCO2eYears);
+        await inventoryValue.save({ transaction });
+
+        activityValue.co2eq = totalCO2e;
+        activityValue.co2eqYears = totalCO2eYears;
```

Comment:
activityValue needs to be saved as well, otherwise this never makes it to the database...

Reply:
```suggestion
        activityValue.co2eq = totalCO2e;
        activityValue.co2eqYears = totalCO2eYears;
        await activityValue.save({ transaction });
```